src/rgw/driver/posix/bucket_cache.h

```diff
@@ -544,6 +550,78 @@ struct BucketCache : public Notifiable
     return rc;
   } /* notify */
   
+  int add_entry(const DoutPrefixProvider* dpp, std::string bname, rgw_bucket_dir_entry bde) {
+    using namespace LMDBSafe;
+
+    GetBucketResult gbr = get_bucket(dpp, bname, BucketCache<D, B>::FLAG_LOCK);
+    auto [b /* BucketCacheEntry */, flags] = gbr;
+    if (b) {
+      unique_lock ulk{b->mtx, std::adopt_lock};
+      ulk.unlock();
+
+      auto txn = b->env->getRWTransaction();
+      auto concat_k = concat_key(bde.key);
+      std::string ser_data;
+      zpp::bits::out out(ser_data);
+      struct timespec ts {
+        ceph::real_clock::to_timespec(bde.meta.mtime)
+      };
+      auto errc =
+          out(bde.key.name, bde.key.instance, /* XXX bde.key.ns, */
+              bde.ver.pool, bde.ver.epoch, bde.exists, bde.meta.category,
+              bde.meta.size, ts.tv_sec, ts.tv_nsec, bde.meta.owner,
+              bde.meta.owner_display_name, bde.meta.accounted_size,
+              bde.meta.storage_class, bde.meta.appendable, bde.meta.etag);
+      if (errc.code != std::errc{0}) {
+        abort();
+      }
+      txn->put(b->dbi, concat_k, ser_data);
+
+      txn->commit();
+      lru.unref(b, cohort::lru::FLAG_NONE);
+    } /* b */
+
+    return 0;
+  } /* add_entry */
+
+  int remove_entry(const DoutPrefixProvider* dpp, std::string bname, cls_rgw_obj_key key) {
+    using namespace LMDBSafe;
+
+    GetBucketResult gbr = get_bucket(dpp, bname, BucketCache<D, B>::FLAG_LOCK);
+    auto [b /* BucketCacheEntry */, flags] = gbr;
+    if (b) {
+      unique_lock ulk{b->mtx, std::adopt_lock};
+      ulk.unlock();
```

Comment:
is it correct unlock before the lmdb txn?

Reply:
(compare l. 620)

Reply:
This is how notify() works, and I copied that.  It gets the lock, unlocks, then creates and commits the transaction.  Maybe the lock isn't necessary for this bit at all?